url433bec1e56_sarobe_DifficultyInvestigation_tgz-pJ8-fr_inria_optimization_cmaes_examples_CMAExample1J8/src/fr/inria/optimization/cmaes/CMAEvolutionStrategy.java

```diff
@@ -0,0 +1,2704 @@
+     *       objective functions preimage, number of variables
+     */
+    public CMAEvolutionStrategy(int dimension) {
+        setDimension(dimension);
+        state = -1;
+    }
+   
+    /** initialization providing all mandatory input arguments at once. The following two 
+     * is equivalent
+     * <PRE>
+          cma.init(N, X, SD);
+     * </PRE> and
+     * <PRE>
+          cma.setInitalX(X);  // 
+          cma.setInitialStandardDeviations(SD);
+          cma.init(N);
+     * </PRE> 
+     *
+     * The call to <code>init</code> is a point of no return for parameter 
+     * settings, and demands all mandatory input be set. <code>init</code> then forces the 
+     * setting up of everything and calls 
+     * <code>parameters.supplementRemainders()</code>. If <code>init</code> was not called before, it is called once in
+     * <code>samplePopulation()</code>. The return value is only provided for sake of convenience. 
+     * 
+     * @param dimension
+     * @param initialX double[] can be of size one, where all variables are set to the 
+     * same value, or of size dimension
+     * @param initialStandardDeviations can be of size one, where all standard
+     * deviations are set to the same value, or of size dimension
+     * 
+     * @return <code>double[] fitness</code> of length population size lambda to assign and pass
+     * objective function values to <code>{@link #updateDistribution(double[])}</code>
+     * 
+     * @see #init()
+     * @see #init(int)
+     * @see #setInitialX(double[])
+     * @see #setTypicalX(double[])
+     * @see #setInitialStandardDeviations(double[])
+     * @see #samplePopulation()
+     * @see CMAParameters#supplementRemainders(int, CMAOptions)
+     */
+    public double[] init(int dimension, double[] initialX, double[] initialStandardDeviations) { 
+    	setInitialX(initialX);
+    	setInitialStandardDeviations(initialStandardDeviations);
+    	return init(dimension);
+    }
+    
+    private double[] getArrayOf(double x, int dim) {
+    	double[] res = new double[dim];
+    	for (int i = 0; i < dim; ++i)
+    		res[i] = x;
+    	return res;
+    }
+    /** 
+     * 
+     * @param x null or x.length==1 or x.length==dim, only for the second case x is expanded
+     * @param dim
+     * @return <code>null</code> or <code>double[] x</code> with <code>x.length==dim</code>
+     */
+    private double[] expandToDimension(double[] x, int dim) {
+    	if (x == null)
+    		return null;
+    	if (x.length == dim)
+    		return x;
+    	if (x.length != 1)
+    		error("x must have length one or length dimension");
+
+    	return getArrayOf(x[0], dim);
+    }
+
+    /**
+     * @param dimension search space dimension 
+     * @see #init(int, double[], double[])
+     * */
+    public double[] init(int dimension) { 
+    	setDimension(dimension);
+    	return init();
+    }
+    /** 
+     * @see #init(int, double[], double[])
+     * */
+    public double[] init() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 42. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 590-595 into a separate method.